Verbs fix: Removed Dependency on Duplicate Recv Flag

yanivbl6 opened the pull request:
This PR is a response for #11825 

I've added a new Queue and Table for the **RdmaTensorBuffer**, where the relevant information from the send-recv waiter table is stored in case the actual rdma-write does not occur (If the allocated buffer is too small, or busy). When we get a RDMA_MESSAGE_BUFFER_RESPONSE, or RDMA_MESSAGE_BUFFER_IDLE, we check the new queue/table instead of checking the waiter table. (This happens in the new ReSendNextItem() function).

Also added a new function that generates the callback lambda function, to avoid code redundancy.
The result is that the code is less clean- I am open to suggestions on making it more tidy.

@junshi15 @shamoya

Commit:
Added new queue and table to the RdmaTensorBuffer, where a semi-replica of the send_recv_ops waiter table is stored.
Values are added to the local table when remote is not ready for receiving our tensor.
When we get an BUFFER_RESPONSE, we use ReSendNextItem instead of sendNextItem.

Splitted SendNextItem And ReSend to two functions, move callback to new function for code reuse

Clanged formatted

Fixed according to shamoya's review, Fixed send->resend flow

squached, rebased on origin/master

Commit:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into removed_dependancy_on_dup_flag_for_verbs

tensorflow-jenkins commented:
Can one of the admins verify this patch?

shamoya commented:
+ @poxvoculi

junshi15 commented on tensorflow/contrib/verbs/rdma.cc:
```diff
@@ -793,9 +841,16 @@ void RdmaTensorBuffer::PostCopyOperations(
+    ReItem* item = new ReItem();
+    item->send_args = send_args;
+    item->recv_args = recv_args;
+    item->in = in;
```
Do we need to increase reference count on [this](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/rendezvous.cc#L217-L219) or [that](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/rendezvous.cc#L175-L177)?
```
      if (item->recv_args.device_context) {
        item->recv_args.device_context->Ref();
      }
```

junshi15 commented:
@yanivbl6, thanks for the quick response. It looks mostly good. I have a comment regarding the reference count.

Commit:
Removed Status from ReItem, added Referencing, added constructor

Added checks for context existaning before Ref()

yanivbl6 commented:
Thanks @junshi15,
I've added a constructor / destructor to the ReItem struct, to take care of the referencing and the de-referencing.  Also removed the unused field (Status).

junshi15 commented:
It looks good to me. @shamoya any comment?

yanivbl6 commented:
@shamoya participated in the code writing, and reviewed it prior to the PR.

poxvoculi commented on tensorflow/contrib/verbs/rdma.h:
```diff
@@ -224,14 +225,52 @@ class RdmaMessageBuffer : public RdmaBuffer {
+    Tensor in;
+    bool is_dead;
+
+    ReItem(const Rendezvous::Args& send_args_, const Rendezvous::Args& recv_args_, const Tensor& in_, bool is_dead_ ): send_args(send_args_), recv_args(recv_args_), in(in_), is_dead(is_dead_){
```
too long line, break by col 80

poxvoculi commented on tensorflow/contrib/verbs/rdma.cc:
```diff
@@ -793,9 +840,12 @@ void RdmaTensorBuffer::PostCopyOperations(
       VLOG(2) << "Extend RDMA buffer from " << size_ << " to " << buffer_size;
     }
     CreateCPUBuffer(buffer_size, false);
+    //Need to be recieved again, put into the re-recv queue and the table
```
s/recieved/received/

Commit:
Clanged reformated, spelling fix

martinwicke commented:
Jenkins, test this please.

yifeif commented:
@tensorflow-jenkins test this please
